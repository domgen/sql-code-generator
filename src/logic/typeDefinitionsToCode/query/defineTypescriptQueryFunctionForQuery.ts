import { camelCase } from 'change-case';
import { castQueryNameToTypescriptTypeName } from '../common/castQueryNameToTypescriptTypeName';

export interface QueryFunctionDefinition {
  code: string;
  imports: {
    queryNameAlias: string;
    generatedTypes: string[];
  };
}

/**
 * given:
 * - typescript type definition of input and output
 *
 * produce:
 * - typescript function that gets the inputs and returns promise of output
 * - typescript function that calls database in best practice way
 *   - uses yesql to pass inputs into parameterized statement
 *   - parses result and extracts output
 *   - logs inputs and output stats
 */
export const defineTypescriptQueryFunctionForQuery = ({ name }: { name: string }): QueryFunctionDefinition => {
  const typescriptTypeName = castQueryNameToTypescriptTypeName({ name });

  // define the query name alias, import { query as ... } under
  const queryNameAlias = `${camelCase(typescriptTypeName)}Sql`; // i.e., sqlQueryUpsertUserSql

  // define the generated types to import
  const outputTypeName = `${typescriptTypeName}Output`;
  const inputTypeName = `${typescriptTypeName}Input`;
  const generatedTypesToImport = [inputTypeName, outputTypeName];

  // define the function body
  const typescriptQueryFunctionName = camelCase(typescriptTypeName); // i.e., same as typename, except camel case
  const queryFunctionDefinition = `
export const ${typescriptQueryFunctionName} = async ({
  dbExecute,
  logDebug,
  input,
}: {
  dbExecute: DatabaseExecuteCommand;
  logDebug: LogMethod;
  input: ${inputTypeName};
}): Promise<${outputTypeName}[]> =>
  executeQueryWithBestPractices({
    dbExecute,
    logDebug,
    name: '${typescriptQueryFunctionName}',
    sql: ${queryNameAlias},
    input,
  });
  `.trim();
  return {
    imports: {
      queryNameAlias,
      generatedTypes: generatedTypesToImport,
    },
    code: queryFunctionDefinition,
  };
};
