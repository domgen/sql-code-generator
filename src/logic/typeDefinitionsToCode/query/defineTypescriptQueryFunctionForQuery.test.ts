import { defineTypescriptQueryFunctionForQuery } from './defineTypescriptQueryFunctionForQuery';

describe('defineTypescriptQueryFunctionForQuery', () => {
  it('should define the method accurately based on name and import path', () => {
    const { code, imports } = defineTypescriptQueryFunctionForQuery({
      name: 'upsert_image',
    });
    expect(imports.generatedTypes.length).toEqual(2);
    expect(imports.generatedTypes).toEqual([
      'SqlQueryUpsertImageInput',
      'SqlQueryUpsertImageOutput',
    ]);
    expect(imports.queryNameAlias).toEqual('sqlQueryUpsertImageSql'); // i.e., import { query as sqlQueryUpsertImageSql } from '../rel/path/to/query';
    expect(code).toMatchSnapshot();
  });
});
