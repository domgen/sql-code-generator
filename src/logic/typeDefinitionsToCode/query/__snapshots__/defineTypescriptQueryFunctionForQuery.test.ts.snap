// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`defineTypescriptQueryFunctionForQuery should define the method accurately based on name and import path 1`] = `
"export const sqlQueryUpsertImage = async ({
  dbExecute,
  logDebug,
  input,
}: {
  dbExecute: DatabaseExecuteCommand;
  logDebug: LogMethod;
  input: SqlQueryUpsertImageInput;
}): Promise<SqlQueryUpsertImageOutput[]> =>
  executeQueryWithBestPractices({
    dbExecute,
    logDebug,
    name: 'sqlQueryUpsertImage',
    sql: sqlQueryUpsertImageSql,
    input,
  });"
`;
